insights/combiners/os_release.py

```diff
@@ -0,0 +1,135 @@
+
+
+@combiner(optional=[InstalledRpms, RedhatRelease, OsRelease, Uname])
+class OSRelease(object):
+    """
+    This combiner can be used to identify if the host an RHEL host by
+    checking the following parsers:
+
+        - :py:class:`insights.parsers.os_release.OsRelease`
+        - :py:class:`insights.parsers.redhat_release.RedhatRelease`
+        - :py:class:`insights.parsers.installed_rpms.InstalledRpms`
+
+    If none of the above parsers is available, NON-RHEL will be returned.
+
+    The :py:class:`insights.parsers.uname.Uname` parser is used to get the
+    booting kernel package.
+    """
+    def __init__(self, rpms, rhr, osr, uname):
+        self._is_rhel = False
+        self._product = "Unknown"
+        if osr:
+            names = list(filter(None,
+                                [osr.get('ID'), osr.get('NAME'),
+                                 osr.get('PRETTY_NAME')]))
+            if names and any(nr in nm.lower() for nm in names
+                                              for nr in NON_RHEL_KEYS):
+                self._product = names[-1]
+                # NON-RHEL Assertion 1: NG /etc/os-release
+                return
+        if rhr and any(nr in rhr.raw.lower() for nr in NON_RHEL_KEYS):
+            self._product = rhr.product
+            # NON-RHEL Assertion 2: NG /etc/redhat-release
+            return
+        if rpms and rpms.packages:
+            self._points = 0
+            self._installed_packages = 1  # kernel is must-installed
+            # check the booting 'kernel' first
+            boot_kn = InstalledRpm.from_package(
+                    'kernel-{0}'.format(uname.kernel)) if uname else None
+            if boot_kn:
+                # check the booting kernel only when Uname is available
+                for pkg in rpms.packages.get('kernel', []):
+                    if pkg == boot_kn:
+                        if (pkg.redhat_signed and
+                                pkg.vendor and pkg.vendor == 'Red Hat, Inc.'):
+                            self._points += 1
+                        break
+            # check other packages
+            for pkg_name in MIN_RHEL_PKGS:
+                if type(pkg_name) is not str:
+                    pkgs = [_p for _p in pkg_name if _p in rpms]
+                    pkg_name = pkgs[0] if pkgs else None
+                pkg = rpms.newest(pkg_name)
+                if pkg:
+                    self._installed_packages += 1
+                    if (pkg.redhat_signed and
+                            pkg.vendor and pkg.vendor == 'Red Hat, Inc.'):
+                        self._points += 1
+            if self._points >= round(THRESHOLD * self._installed_packages):
+                # RHEL: more than THRESHOLD packages are from Red Hat
+                self._is_rhel = True
+                self._product = RHEL_STR
+        elif ((rhr and rhr.is_rhel) or
+                (osr and any(k in v.lower() for k in RHEL_KEYS for v in names))):
```

Comment:
I'm not sure if I'm missing it or not, but could names be accessed before being defined? It's only defined if osr exist, but that's not checked before trying to loop through names.

Reply:
@ryan-blakley - Thanks for reviewing.
from the following `if`
```python
        if osr:
            names = list(filter(None,
                                [osr.get('ID'), osr.get('NAME'),
                                 osr.get('PRETTY_NAME')]))
            if ...:
                  return
```
it ensures the `names` is defined when the `osr` is available
and this check `(osr and any(k in v.lower() for k in RHEL_KEYS for v in names))` guaranteed it will be checked only when `osr` is available.

But I think it doesn't hurt to modify it as `(osr and names and any(k in v.lower() for k in RHEL_KEYS for v in names))` which looks more friendly as well.

Reply:
Updated